### PR TITLE
[WIP] Enable TF32 support for cuDNN

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39361 [WIP] Enable TF32 support for cuDNN**
* #39283 [WIP] Enable TF32 support for cuBLAS

### CODE REVIEW

File: aten/src/ATen/native/cudnn/Conv.cpp [hunk 1]
```diff
@@ -657,7 +657,13 @@ class AlgoIterator {
     if (args.params.dataType == CUDNN_DATA_HALF) {
       perfResults[0].mathType = CUDNN_TENSOR_OP_MATH;
     } else {
+#if defined(CUDNN_VERSION) && CUDNN_VERSION >= 8000
+      if (!at::globalContext().useTF32CuDNN()) {
+        perfResults[0].mathType = CUDNN_FMA_MATH;
+      }
+#else
       perfResults[0].mathType = CUDNN_DEFAULT_MATH;
+#endif
```
Comment (mcarilli) [inline aten/src/ATen/native/cudnn/Conv.cpp:666]: The above set of conditionals won't touch `mathType` at all `#if defined(CUDNN_VERSION) && CUDNN_VERSION >= 8000` and `at::globalContext().useTF32CuDNN()` is true.  If `mathType` is `CUDNN_DEFAULT_MATH` by default (which would make sense to me...) maybe that's fine, but maybe we want to ensure it's explicitly set for all possible compile and runtime options, ie
```
#if defined(CUDNN_VERSION) && CUDNN_VERSION >= 8000
      if (!at::globalContext().useTF32CuDNN()) {
        perfResults[0].mathType = CUDNN_FMA_MATH;
      } else
#else
      {
        perfResults[0].mathType = CUDNN_DEFAULT_MATH;
      }
#endif
```

Comment (zasdfgbnm) [reply aten/src/ATen/native/cudnn/Conv.cpp:667]: I believe this is resolved

**Comment by zasdfgbnm** [comment]
I messed up my stack during a rebase, so this is gone. I will open a new PR when we are ready to merge.

**Comment by zasdfgbnm** [comment]
[Deprecated]

**Pull request closed without merging**